### PR TITLE
Dialog: Full screen & Bottom sheet support

### PR DESCRIPTION
**Pull request by maximedegreve**
### Closes
- https://github.com/github/primer/issues/2414
- https://github.com/github/primer/issues/2413

Note: I've chosen not to tackle the[ issues mentioned here](https://github.com/github/primer/issues/2372) to avoid causing any disruptive changes.

#### New

1️⃣ `action-sheet` dialog

We've implemented a range slider to easily adjust the size of the dialog using only a keyboard, enhancing accessibility. We've used a similar approach for our [SplitPageLayout resizer](https://github.com/primer/react/pull/3903). I've also implemented reduced-motion support.

https://github.com/primer/react/assets/980622/c55fc0cb-1cea-4dab-94cb-e9239a870431

https://github.com/primer/react/assets/980622/babadea0-8794-40b1-a812-cf9b6f7617f8

2️⃣  `full-screen` dialog

This is quite simple, as it closely resembles the existing dialog behavior, with the only difference being that it's full-screen.

https://github.com/primer/react/assets/980622/eb043543-8a06-4854-bf0b-fc6c446213d2

#### Changed

- Storybook features always display dialogs in an open state so that snapshots can effectively capture any changes.
- Removed the ability to change props in features for simplicity and consistency with other components
- In the example called 'WithCustomRenderers', I've placed the header, content, and footer components inside the 'WithCustomRenderers' component. This change improves readability, especially when highlighting the code.

### Rollout strategy

- [ ] Patch release
- [x] Minor release
- [ ] Major release; if selected, include a written rollout or migration plan
- [ ] None; if selected, include a brief description as to why

### Merge checklist

- [ ] Added/updated tests
- [x] Added/updated documentation
- [x] Added/updated previews (Storybook)
- [x] Changes are [SSR compatible](https://github.com/primer/react/blob/main/contributor-docs/CONTRIBUTING.md#ssr-compatibility)
- [x] Tested in Chrome
- [x] Tested in Firefox
- [x] Tested in Safari
- [x] Tested in Edge
- [ ] (GitHub staff only) Integration tests pass at github/github ([Learn more about how to run integration tests](https://github.com/github/primer-engineering/blob/main/how-we-work/testing-primer-react-pr-at-dotcom.md))

### CODE REVIEW

**Comment by joshblack** [comment]
Hey @maximedegreve! Sorry for the delay on this, was spending some time reading the related issues to get the context to understand the changes 👀 

Just wanted to check-in with you to make sure I understood things right, the corresponding issues are talking about two things: 

1. The ability for a dialog to go full screen (the `full-screen` type in this PR)
2. A new presentation for a dialog called an action sheet (the `action-sheet` type in this PR)

And it looks we're supporting using these types responsively through the `type` prop on `Dialog` to get something like:

```tsx
<Dialog type={['full-screen', 'default']} />
```

Hope that makes sense 🤔 Let me know if I'm missing anything.

With this in mind, I had a couple of questions come up:

- Are these new types restricted to any specific viewport sizes or should they be visible across viewports?
- Does action sheet in this PR correspond to bottom sheets in the design guidance or are they a separate type?

**Comment by joshblack** [comment]
@maximedegreve nice, makes sense. Would it be helpful to limit usage of these variants to only specific breakpoints (if feasible) so that these are only used in appropriate contexts? Or even just roll these into the components themselves as default behavior? (This would like require a specific component for BottomSheet but could be worth it)

**Comment by maximedegreve** [comment]
@joshblack 💯 feasible. At first, I thought the same. But I have concerns about limiting it like this. Some dialogs already are playing with a full screen mode right now in desktop and bottom sheets are also very similar to side sheets. I would assume side sheets would be an addition to the future. Will our API become inconsistent and harder to read if we allow side sheets on desktop but not bottom sheets? In this case we would need `type: sidesheet bottomsheetOnMobile: true` or `type:default bottomSheetOnMobile:true`.  In all our other components we use the {narrow: value, default: value...} pattern.

Red arrow pointing to a full screen icon a 'New conversation' header.

I do believe that components using the dialog, like SelectPanel, might prefer these built-in defaults though. 

It could be a good idea to check with the rest of team if we want to maintain flexibility at the component level here or not.

File: src/ConfirmationDialog/ConfirmationDialog.tsx
Comment (mperrotti) [inline src/ConfirmationDialog/ConfirmationDialog.tsx:20]: Would we ever even use an action sheet or full screen confirmation dialog?

File: src/Dialog/Dialog.docs.json
Comment (mperrotti) [inline src/Dialog/Dialog.docs.json:50]: I'm wondering if we should just make a `'bottom-sheet'` dialog it's own component instead of rendering it using a `Dialog`. Then, we could set `'full-screen'` in the `width` and `height` props. If `'full-screen'` is passed to _either_ `width` or `height`, the dialog would be full screen.

Curious to hear what other reviewers think.

File: src/Dialog/Dialog.tsx
Comment (mperrotti) [inline src/Dialog/Dialog.tsx:111]: Some concern as I posted in `ConfirmationDialog`:

> It could be confusing if somebody tries to pass 'drag' or 'overlay' on something that doesn't use those gestures. I'm not sure of a better way to handle those, but maybe we can explicitly mention which variants accept which gestures in the JSDoc comment and prop docs.

File: src/Dialog/Dialog.tsx
Comment (mperrotti) [inline src/Dialog/Dialog.tsx:507]: What happens if we don't set a `pointer-events` style at all? I'd assume it would be the same as `pointer-events: auto;` unless we style `pointer-events` somewhere else that I missed.

File: src/Dialog/DialogBottomSheet.tsx
Comment (mperrotti) [inline src/Dialog/DialogBottomSheet.tsx:22]: Non-blocking nitpick: I'd reword this to "Content that appears in the draggable header area." The current description kind of sounds like it lets you change what we render instead of the the blue drag "handle".

File: src/Dialog/DialogBottomSheet.tsx
Comment (mperrotti) [inline src/Dialog/DialogBottomSheet.tsx:46]: This seems right to me, but I'd double-check with a11y folks. Some of the `aria-` attributes have inconsistent behavior on different screen readers.

File: src/Dialog/DialogBottomSheet.tsx
Comment (mperrotti) [inline src/Dialog/DialogBottomSheet.tsx:105]: Non-blocking nitpick: `isReduced` should already be in scope and shouldn't need to be redefined. I don't think it would change between render and `onDragStop`.

File: src/internal/components/ScrollableRegion.tsx
Comment (mperrotti) [inline src/internal/components/ScrollableRegion.tsx:10]: I think I'm missing something. Why do we need to start supporting `sx` on `ScrollableRegion`?

File: src/Dialog/DialogBottomSheet.tsx
Comment (mperrotti) [inline src/Dialog/DialogBottomSheet.tsx:201]: Neat technique!

File: src/Dialog/DialogBottomSheet.tsx
Comment (mperrotti) [inline src/Dialog/DialogBottomSheet.tsx:197]: I think it might be a little nicer if the close button was the first focusable element in the DOM. This way, arrow buttons don't move the sheet around until I explicitly focus the drag handle.

Comment (mperrotti) [inline src/Dialog/DialogBottomSheet.tsx:197]: The drag target is layered above the close button, so it isn't clickable.

File: src/Dialog/DialogBottomSheet.tsx
Comment (mperrotti) [inline src/Dialog/DialogBottomSheet.tsx:92]: We might get better performance if we always render it at full height and use `translate: transformY();` to position it in and out of the viewport.

Changing `height` triggers reflow (and maybe repaint?).

**Comment by mperrotti** [comment]
Also, we need to add tests: unit tests (if applicable) and visual regression tests.

File: src/Dialog/Dialog.docs.json
Comment (maximedegreve) [reply src/Dialog/Dialog.docs.json:50]: My concern is that you'll need to wrap different components in a `<Box/>` for responsive behaviour. For instance, we're planning to integrate side sheets shortly too. In this scenario, I'd prefer using a side sheet on desktop and a bottom sheet on mobile. This means I'll be required to proceed in this manner:

```javascript
<Box sx={{display: ['block', 'block', 'none']}}><BottomSheet /></Box>
<Box sx={{display: ['none', 'none', 'block']}}><SideSheet /></Box>
``` 

Another situation where I need to follow this approach is when I want to display a bottom sheet on mobile, but a standard dialog on desktop.

```javascript
<Box sx={{display: ['block', 'block', 'none']}}><BottomSheet /></Box>
<Box sx={{display: ['none', 'none', 'block']}}><Dialog /></Box>
``` 

This adds extra complexity because an open `<Dialog/>` that turns into `<BottomSheet/>` on mobile would need to stay open. If the user applies separate states for each component, it might lead to unintended effects.

The proposed API really simplifies above and ensures consumers still understand the underlying semantics is a dialog.

```javascript
<Dialog title="Your title" type={{
    narrow: 'full-screen',
    regular: 'default',
    wide: 'default'
}} >
```

or

```javascript
<Dialog title="Your title" type={{
    narrow: 'bottom-sheet',
    regular: 'default',
    wide: 'default'
}} >
```

File: src/Dialog/Dialog.tsx
Comment (maximedegreve) [reply src/Dialog/Dialog.tsx]: Sadly if I remove it doesn't work anymore.

File: src/internal/components/ScrollableRegion.tsx
Comment (maximedegreve) [reply src/internal/components/ScrollableRegion.tsx:10]: Yes, because if we don't, the footers in the bottom sheets and full screens won't stay at the bottom if the content isn't long enough. That's why I'm using this method:

`<ScrollableRegion aria-labelledby={dialogLabelId} className="DialogOverflowWrapper" sx={{flex: 1}}>`

File: src/Dialog/Dialog.tsx
Comment (maximedegreve) [reply src/Dialog/Dialog.tsx:111]: I believe it's clearer in this context because the dialogue supports it with the bottom sheet variant. But I'm curious if those gestures are currently in used. Why would you need to know how the user closed the dialogue? What do you think...

File: src/Dialog/DialogBottomSheet.tsx
Comment (maximedegreve) [reply src/Dialog/DialogBottomSheet.tsx:197]: Fixed

File: src/ConfirmationDialog/ConfirmationDialog.tsx
Comment (maximedegreve) [reply src/ConfirmationDialog/ConfirmationDialog.tsx]: Fixed

File: src/Dialog/DialogBottomSheet.tsx
Comment (maximedegreve) [reply src/Dialog/DialogBottomSheet.tsx]: Makes sense. Fixed.

File: src/Dialog/DialogBottomSheet.tsx
Comment (maximedegreve) [reply src/Dialog/DialogBottomSheet.tsx:46]: @ichelsea @ericwbailey is that something you could double check for me?

File: src/Dialog/DialogBottomSheet.tsx
Comment (maximedegreve) [reply src/Dialog/DialogBottomSheet.tsx]: Whoops yes this is overkill. Fixed.

File: src/Dialog/DialogBottomSheet.tsx
Comment (maximedegreve) [reply src/Dialog/DialogBottomSheet.tsx:92]: Are you sure it will reflow/repaint because the content is scrollable therefore the width and height of the scrollable content won't change. I haven't observed any performance issues. Not animating the height is kinda complex because the scrollable area needs to resize either way. Using `transform` I would have to set a height before animating to something beyond the height of the viewport and at the end of the animation then set the final height to ensure the scrollable area is correct. Things get even more tricky while you're dragging.

File: src/Dialog/DialogBottomSheet.tsx
Comment (maximedegreve) [reply src/Dialog/DialogBottomSheet.tsx:196]: Fixed.

File: src/Dialog/Dialog.tsx
Comment (keithamus) [reply src/Dialog/Dialog.tsx:111]: We should avoid extending these. The gestures aren't checked in consuming code, and in the future the `onClose()` call will be called whenever the dialog is closed and will not pass a gesture back. See https://github.com/primer/react/pull/3840/files#diff-7c9e4f2540e063c93f543e6eaa44cac7dc3b0e55a8fe81030c9e571f72950a3eR97-R102

File: src/Dialog/DialogBottomSheet.tsx
Comment (mperrotti) [reply src/Dialog/DialogBottomSheet.tsx:46]: Also tagging @alexislucio since she's this week's a11y FR.

File: src/Dialog/DialogBottomSheet.tsx
Comment (mperrotti) [reply src/Dialog/DialogBottomSheet.tsx:92]: I checked it out in Chrome's performance dev tools, and it seems to behave just fine on my work computer. Let's leave this as-is unless we notice jank on mobile or other less powerful devices. 🙂 

---

> Are you sure it will reflow/repaint because the content is scrollable therefore the width and height of the scrollable content won't change.

I'm pretty sure the browser does a reflow whenever you change a height or width, even if no content is actually affected. Browsers even trigger a reflow when you check the dimensions of an element using `.getBoundingClientRect()` 😢

**Review by mperrotti — COMMENTED** [review]
Looking good to me! There's one other thing I noticed in [this story](https://primer-bfeba1f965-13348165.drafts.github.io/storybook/?path=/story/components-dialog-features--bottom-sheet): the dialog doesn't dismiss when I click the close button, but it _does_ dismiss when I click the backdrop.

File: src/Dialog/DialogBottomSheet.tsx
Comment (maximedegreve) [reply src/Dialog/DialogBottomSheet.tsx:92]: Oh right. I've learned something today. 🧠

File: src/Dialog/Dialog.tsx
Comment (maximedegreve) [reply src/Dialog/Dialog.tsx]: @keithamus Makes a lot of sense. I've removed it and always return `close-button` for now.

File: src/Dialog/Dialog.tsx
Comment (keithamus) [inline src/Dialog/Dialog.tsx:301]: There is a lot of duplication going on here. It might benefit from hoisting the element into a variable. Something like:


```js
let TheDialog = NormalDialog;
if (responsiveType === 'full-screen') TheDialog = FullScreenDialog
else if (responsiveType === 'bottom-sheet') TheDialog = BottomSheetDialog
```

I haven't looked at the duplication within each of these components but it might be worth looking into that too.

File: src/Dialog/DialogBottomSheet.tsx
Comment (keithamus) [inline src/Dialog/DialogBottomSheet.tsx:297]:
```suggestion
const Content = styled.dialog<
```

File: src/Dialog/DialogBottomSheet.tsx
Comment (keithamus) [inline src/Dialog/DialogBottomSheet.tsx:177]: Why does this container exist?

File: src/__tests__/Dialogv2.test.tsx
Comment (mperrotti) [inline src/__tests__/Dialogv2.test.tsx:58]: We don't do snapshot tests anymore because the errors potentially caught in snapshots are caught by VRTs.

It looks like we never had any unit tests for Dialog v2, so I don't think you should have to add them in this PR.

File: src/Dialog/DialogBottomSheet.tsx
Comment (mperrotti) [inline src/Dialog/DialogBottomSheet.tsx:1]: We should add unit tests for this component. We would test things like whether the `onClose` prop is called when passed. You can use `npm run test:coverage -- src/Dialog/DialogBottomSheet.tsx` to get ideas for what we might want to test for.

File: src/Dialog/DialogBottomSheet.tsx
Comment (maximedegreve) [reply src/Dialog/DialogBottomSheet.tsx:177]: It doesn't have to, but there are subtle differences that cause them to behave differently. When animation is used in the code, it can be easier to understand if you organize the different states by their names. Of course, this is a matter of personal preference, and you could also consider adding an 'isFullScreen' boolean property here.

``` JavaScript
const NormalDialog = styled.div<StyledDialogProps>`
  ...
  box-shadow: ${get('shadows.overlay.shadow')};
  min-width: 296px;
  max-width: calc(100vw - 64px);
  max-height: calc(100vh - 64px);
  ...
  width: ${props => widthMap[props.width ?? ('xlarge' as const)]};
  height: ${props => heightMap[props.height ?? ('auto' as const)]};
  border-radius: 12px;
  ...
  @keyframes overlay--dialog-appear {
    0% {
      opacity: 0;
      transform: scale(0.5);
    }
    100% {
      opacity: 1;
      transform: scale(1);
    }
  }

  ${sx};
`

const FullScreenDialog = styled.div<StyledDialogProps>`
  ...
  width: 100vw;
  height: 100vh;
  ...
  animation: overlay--dialog-appear ${ANIMATION_DURATION} ${get('animation.easeOutCubic')};
  @keyframes overlay--dialog-appear {
    0% {
      opacity: 0;
    }
    100% {
      opacity: 1;
    }
  }
  ${sx};
`
```

File: src/Dialog/Dialog.tsx
Comment (maximedegreve) [reply src/Dialog/Dialog.tsx:301]: It doesn't have to, but there are subtle differences that cause them to behave differently. When animation is used in the code, it can be easier to understand if you organize the different states by their names. Of course, this is a matter of personal preference, and you could also consider adding an 'isFullScreen' boolean property here.

``` JavaScript
const NormalDialog = styled.div<StyledDialogProps>`
  ...
  box-shadow: ${get('shadows.overlay.shadow')};
  min-width: 296px;
  max-width: calc(100vw - 64px);
  max-height: calc(100vh - 64px);
  ...
  width: ${props => widthMap[props.width ?? ('xlarge' as const)]};
  height: ${props => heightMap[props.height ?? ('auto' as const)]};
  border-radius: 12px;
  ...
  @keyframes overlay--dialog-appear {
    0% {
      opacity: 0;
      transform: scale(0.5);
    }
    100% {
      opacity: 1;
      transform: scale(1);
    }
  }

  ${sx};
`

const FullScreenDialog = styled.div<StyledDialogProps>`
  ...
  width: 100vw;
  height: 100vh;
  ...
  animation: overlay--dialog-appear ${ANIMATION_DURATION} ${get('animation.easeOutCubic')};
  @keyframes overlay--dialog-appear {
    0% {
      opacity: 0;
    }
    100% {
      opacity: 1;
    }
  }
  ${sx};
`
```

File: src/Dialog/DialogBottomSheet.tsx
Comment (maximedegreve) [reply src/Dialog/DialogBottomSheet.tsx:177]: Same answer as: https://github.com/primer/react/pull/4048#discussion_r1455511524

File: src/__tests__/Dialogv2.test.tsx
Comment (maximedegreve) [reply src/__tests__/Dialogv2.test.tsx:58]: That makes life much easier!

File: src/Dialog/DialogBottomSheet.tsx
Comment (mperrotti) [reply src/Dialog/DialogBottomSheet.tsx:1]: I can try and help too. Just push up your changes.

Comment (maximedegreve) [reply src/Dialog/DialogBottomSheet.tsx:1]: Done

**Comment by mperrotti** [comment]
I'm working on fixing your Jest issues, but it also looks like some styles broke.

Screenshot 2024-01-17 at 3 18 06 PM

File: src/Dialog/Dialog.tsx
Comment (mperrotti) [inline src/Dialog/Dialog.tsx:456]: Is this meant to be a `dialog` instead of a `div`?

File: src/Dialog/DialogBottomSheet.test.tsx
Comment (mperrotti) [inline src/Dialog/DialogBottomSheet.test.tsx:15]: If you plan to use `matchMedia` to test things at different media queries, you probably want to change this to `afterEach`

File: src/Dialog/DialogBottomSheet.test.tsx
Comment (mperrotti) [inline src/Dialog/DialogBottomSheet.test.tsx:30]: I don't think this has a default export.

```suggestion
    default: undefined,
```

File: src/Dialog/Dialog.tsx
Comment (mperrotti) [reply src/Dialog/Dialog.tsx:456]: If these _are_ meant to be `dialog`s, we'll need to update the ref types to `HTMLDialogComponent` like I suggested in `src/Dialog/DialogBottomSheet.tsx`

**Review by mperrotti — COMMENTED** [review]
I played around with this and I think I have a way forward (or at least a workaround) to resolve your Jest issues. Maybe we can pair? Ideally with another PRC engineer present 😅

File: src/Dialog/Dialog.tsx
Comment (maximedegreve) [reply src/Dialog/Dialog.tsx:428]: Yes whoops!

File: src/Dialog/DialogBottomSheet.tsx
Comment (maximedegreve) [reply src/Dialog/DialogBottomSheet.tsx]: Actually I had to revert this again because the role is defined through the `role` prop.

File: src/Dialog/DialogBottomSheet.tsx
Comment (maximedegreve) [reply src/Dialog/DialogBottomSheet.tsx]: Actually I had to revert this again because the role is defined through the `role` prop.

File: src/Dialog/DialogBottomSheet.tsx
Comment (maximedegreve) [reply src/Dialog/DialogBottomSheet.tsx]: Actually I had to revert this again because the role is defined through the `role` prop.

File: src/Dialog/Dialog.tsx
Comment (maximedegreve) [reply src/Dialog/Dialog.tsx]: Actually I had to revert this again because the role is defined through the `role` prop.

File: src/Dialog/DialogBottomSheet.test.tsx
Comment (mperrotti) [inline src/Dialog/DialogBottomSheet.test.tsx:33]: If this merges before https://github.com/primer/react/pull/4048, I mighjt update my PR to move these tests into the same test file.

**Review by mperrotti — COMMENTED** [review]
It's looking good! Just one little thing we might to address: we don't delay calling `onClose` when clicking the "x" or keying <kbd>Esc</kbd>

https://github.com/primer/react/assets/2313998/cf97acba-38ca-4298-a222-01a69f23a813

If this is intentional, don't worry about it.

**Comment by maximedegreve** [comment]
Yeah that was intentionally because I thought you aren't swiping it away.

**Review by mperrotti — APPROVED** [review]
Looking good. Since this is such a big PR, it might be nice to get a review from somebody who's "officially" and engineer before merging. Not required though 🙂

**Pull request merged**
